### PR TITLE
Add isolated charged hadron flag to RECO+AOD

### PR DESCRIPTION
**Pull request by ahinzmann**
Followup of #18308

This PR adds a valueMap "chargedHadronPFTrackIsolation" for the PFCandisdates to RECO+AOD.
Charged hadrons with track pT > 1 GeV and rawCaloEnergy > 0.5 GeV are tagged on RECO.
With isolated charged hadrons, calorimeter response can be measured in data and simulation with fine granularity, enabling PF hadron calibration at AOD level.

runTheMatrix -l limited works.

It was checked that the variables in RECO+AOD are filled correctly with reasonable values.

Size with/without this change based on workflow 10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017 is summarized below:

New collection in RECO+AOD:
booledmValueMap_chargedHadronPFTrackIsolation__RECO. 677.8 278.7

### CODE REVIEW

**Comment by perrotta** [comment]
@ahinzmann : is this supposed to replace #18308 ?
(Please clarify, and close any of them which is not needed any more)

**Comment by ahinzmann** [comment]
#18308 is factorized in two PR. This is supposed to replace the AOD+RECO part of #18308. MiniAOD will follow after discussion at xpog

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
On 4/12/17 1:51 AM, ahinzmann wrote:
> Followup of #18308 <https://github.com/cms-sw/cmssw/pull/18308>
> 

what's the reason to move to a new PR?
This disrupts review process.

**Comment by ahinzmann** [comment]
Renamed according to comment by @slava77 in #18308

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFProducer/python/chargedHadronPFTrackIsolation_cfi.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import FWCore.ParameterSet.Config as cms
+
+
+chargedHadronPFTrackIsolation = cms.EDProducer(
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/python/chargedHadronPFTrackIsolation_cfi.py]: this file is redundant and can be removed.
It is already provided by the framework with the call ``` descriptions.add("chargedHadronPFTrackIsolation", desc); ``` available in ChargedHadronPFTrackIsolationProducer

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidateFwd.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/ParticleFlowReco/interface/PFBlock.h"
+#include "DataFormats/ParticleFlowReco/interface/PFBlockElement.h"
+
+namespace edm {
+   class ConfigurationDescriptions;
+}
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: this forward declaration can be removed (there is an include for this class earlier in this file)

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+class ChargedHadronPFTrackIsolationProducer : public edm::stream::EDProducer<>
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: this module can be made ```edm::global```

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+   class ConfigurationDescriptions;
+}
+
+class ChargedHadronPFTrackIsolationProducer : public edm::stream::EDProducer<> 
+{
+public:
+
+  explicit ChargedHadronPFTrackIsolationProducer(const edm::ParameterSet& cfg);
+  ~ChargedHadronPFTrackIsolationProducer() {}
+  void produce(edm::Event& evt, const edm::EventSetup& es);
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: please add ```override ``` at the end

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  static void fillDescriptions(edm::ConfigurationDescriptions & descriptions);
+
+private:
+  // input collection
+  edm::InputTag srcCandidates_;
+  edm::EDGetTokenT<edm::View<reco::PFCandidate> > Candidates_token;
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: please keep the same style of local members: start with lower case, end with an underscore

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  
+  produces<edm::ValueMap<bool> >();
+}
+
+void ChargedHadronPFTrackIsolationProducer::produce(edm::Event& evt, const edm::EventSetup& es) 
+{
+  // get a view of our Candidates via the base candidates
+  typedef edm::View<reco::PFCandidate> PFCandidateView;
+  edm::Handle<PFCandidateView> Candidates;
+  evt.getByToken(Candidates_token, Candidates);
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: start local variables and data members with lower case letter. Capitalization is reserved for class names.

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  
+  std::vector<bool> values;
+
+  for( PFCandidateView::const_iterator c = Candidates->begin(); c!=Candidates->end(); ++c) {
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: for range loop seems more appropriate ``` for (auto cons& c : *candidates) { ```

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    // Check that there is only one track in the block.
+    unsigned int nTracks = 0;
+    if ((c->particleId()==1) && (c->pt()>minTrackPt_) && ((c->rawEcalEnergy()+c->rawHcalEnergy())>minRawCaloEnergy_)) {
+      const reco::PFCandidate::ElementsInBlocks& theElements = c->elementsInBlocks();
+      if( theElements.empty() ) continue;
+      const reco::PFBlockRef blockRef = theElements[0].first;
+      const edm::OwnVector<reco::PFBlockElement>& elements = blockRef->elements();
+      for(unsigned int iEle=0; iEle<elements.size(); iEle++) {	         	 // Find the tracks in the block
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: similarly 
```c++
 for (auto const& ele : elements){
...
```

File: RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc (hunk 3 above)
Comment (ahinzmann) [inline RecoParticleFlow/PFProducer/plugins/ChargedHadronPFTrackIsolationProducer.cc]: made edm::global and changed the produce method to "void produce(edm::StreamID, edm::Event& evt, const edm::EventSetup& es) const" as defined in global/EDProducerBase.h.

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/18324	62fd67c
- new producer to make ``` booledmValueMap_chargedHadronPFTrackIsolation``` is introduced as described; the branch appears up to AOD.
- jenkins tests pass and AOD ouput in 1000.0 shows the new branch is in place
- local test with 136.76 shows that the module is running at somewhat negligible CPU cost (0.5 ms/evt) and creates an output of about 100 bytes/event in size (one bool per about 2K PF candidates in this sample with about 50:1 false:true ratio).

**Comment by davidlange6** [comment]
+1

**Pull request merged**